<a href="https://colab.research.google.com/github/yashasangani/Application-Programming-Interface/blob/main/HTTP_requests_and_Libraries.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import json
import requests

HTTP requests- Web API

GET Method:
1. GET method is used to fetch data from API.
2. It is safe and idempotent method, meaning it can be called multiple times without changing the state of the server.
3. GET requests are typically used to retrieve the information or fetch resources or retrieve data from dataset.

In [ ]:
get_response = requests.get('https://dummyjson.com/products')
get_response.status_code

200

In [ ]:
get_response.text

'{"products":[{"id":1,"title":"Essence Mascara Lash Princess","description":"The Essence Mascara Lash Princess is a popular mascara known for its volumizing and lengthening effects. Achieve dramatic lashes with this long-lasting and cruelty-free formula.","category":"beauty","price":9.99,"discountPercentage":7.17,"rating":4.94,"stock":5,"tags":["beauty","mascara"],"brand":"Essence","sku":"RCH45Q1A","weight":2,"dimensions":{"width":23.17,"height":14.43,"depth":28.01},"warrantyInformation":"1 month warranty","shippingInformation":"Ships in 1 month","availabilityStatus":"Low Stock","reviews":[{"rating":2,"comment":"Very unhappy with my purchase!","date":"2024-05-23T08:56:21.618Z","reviewerName":"John Doe","reviewerEmail":"john.doe@x.dummyjson.com"},{"rating":2,"comment":"Not as described!","date":"2024-05-23T08:56:21.618Z","reviewerName":"Nolan Gonzalez","reviewerEmail":"nolan.gonzalez@x.dummyjson.com"},{"rating":5,"comment":"Very satisfied!","date":"2024-05-23T08:56:21.618Z","reviewerNam

In [ ]:
get_response.json()['products'][0]

{'id': 1,
 'title': 'Essence Mascara Lash Princess',
 'description': 'The Essence Mascara Lash Princess is a popular mascara known for its volumizing and lengthening effects. Achieve dramatic lashes with this long-lasting and cruelty-free formula.',
 'category': 'beauty',
 'price': 9.99,
 'discountPercentage': 7.17,
 'rating': 4.94,
 'stock': 5,
 'tags': ['beauty', 'mascara'],
 'brand': 'Essence',
 'sku': 'RCH45Q1A',
 'weight': 2,
 'dimensions': {'width': 23.17, 'height': 14.43, 'depth': 28.01},
 'warrantyInformation': '1 month warranty',
 'shippingInformation': 'Ships in 1 month',
 'availabilityStatus': 'Low Stock',
 'reviews': [{'rating': 2,
   'comment': 'Very unhappy with my purchase!',
   'date': '2024-05-23T08:56:21.618Z',
   'reviewerName': 'John Doe',
   'reviewerEmail': 'john.doe@x.dummyjson.com'},
  {'rating': 2,
   'comment': 'Not as described!',
   'date': '2024-05-23T08:56:21.618Z',
   'reviewerName': 'Nolan Gonzalez',
   'reviewerEmail': 'nolan.gonzalez@x.dummyjson.com'},

In [ ]:
get_response.json()['products'][0]['title']

'Essence Mascara Lash Princess'

POST Method:
1. the POST method is used to submit data to the server
2. it is commonly use to create new resources or send data to proceed.
3.POST requests contains the request body that hold the data to be submitted.

In [ ]:
headers = { 'Content-Type': 'application/json' }  # Indicates the type of data being sent to the server
body = {'title': 'BMW Pencil'}
string_body = json.dumps(body)
post_response = requests.post('https://dummyjson.com/products/add',headers = headers,data = string_body)
post_response.status_code

201

In [ ]:
post_response.json()

{'id': 195, 'title': 'BMW Pencil'}

In [ ]:
dict_example = [
    {"name": "jhon", "occupation": "firefighter"},
    {"name": "alex", "occupation": "nurse"},
    {"name": "joe", "occupation": "food tester"}
]
type(dict_example)

list

In [ ]:
print(json.dumps(dict_example, indent = 3))

[
   {
      "name": "jhon",
      "occupation": "firefighter"
   },
   {
      "name": "alex",
      "occupation": "nurse"
   },
   {
      "name": "joe",
      "occupation": "food tester"
   }
]


# Problem statement
Write a function that interacts with the 'Restful Booker' API to retrieve guest booking details for a specified range of booking IDs. The returned data will enable analysts to efficiently analyze and process hotel reservation information.
# New Section
API Details: https://restful-booker.herokuapp.com/apidoc/index.html.
# Input:
The function should take two integer inputs: a starting booking ID (start_booking_id) and an ending booking ID (start_booking_id).To performs this task for a specified range of booking IDs (from 1 to 9).
# Output:
[{'firstname': 'Mary', 'lastname': 'Ericsson', 'totalprice': 508, 'depositpaid': False, 'bookingdates': {'checkin': '2016-01-18', 'checkout': '2020-12-23'}, 'additionalneeds': 'Breakfast'}, {'firstname': 'Susan', 'lastname': 'Jones', 'totalprice': 448, 'depositpaid': False, 'bookingdates': {'checkin': '2020-08-02', 'checkout': '2023-02-05'}}, {'firstname': 'Mary', 'lastname': 'Jackson', 'totalprice': 698, 'depositpaid': True, 'bookingdates': {'checkin': '2015-11-17', 'checkout': '2016-10-25'}}, {'firstname': 'Eric', 'lastname': 'Smith', 'totalprice': 922, 'depositpaid': False, 'bookingdates': {'checkin': '2015-02-08', 'checkout': '2021-07-26'}, 'additionalneeds': 'Breakfast'}, {'firstname': 'Mary', 'lastname': 'Jones', 'totalprice': 665, 'depositpaid': False, 'bookingdates': {'checkin': '2023-09-03', 'checkout': '2023-09-05'}}, {'firstname': 'Jim', 'lastname': 'Jackson', 'totalprice': 319, 'depositpaid': False, 'bookingdates': {'checkin': '2021-07-01', 'checkout': '2023-06-28'}, 'additionalneeds': 'Breakfast'}, {'firstname': 'Jim', 'lastname': 'Ericsson', 'totalprice': 160, 'depositpaid': False, 'bookingdates': {'checkin': '2019-11-18', 'checkout': '2020-03-04'}}, {'firstname': 'Mary', 'lastname': 'Wilson', 'totalprice': 503, 'depositpaid': False, 'bookingdates': {'checkin': '2018-01-23', 'checkout': '2021-08-04'}}, {'firstname': 'Mary', 'lastname': 'Jones', 'totalprice': 217, 'depositpaid': True, 'bookingdates': {'checkin': '2015-06-04', 'checkout': '2017-08-06'}}]
# Hint:
Utilize the GET /booking/{bookingId} endpoint to retrieve the details of each booking.

In [13]:
import requests
import json
response = requests.get('https://restful-booker.herokuapp.com/booking')
response.status_code

200

In [14]:
response.text

'[{"bookingid":3120},{"bookingid":3542},{"bookingid":2009},{"bookingid":2130},{"bookingid":2722},{"bookingid":181},{"bookingid":3655},{"bookingid":3799},{"bookingid":797},{"bookingid":3910},{"bookingid":3894},{"bookingid":3897},{"bookingid":2069},{"bookingid":2741},{"bookingid":1679},{"bookingid":3845},{"bookingid":2115},{"bookingid":2386},{"bookingid":3000},{"bookingid":1166},{"bookingid":2765},{"bookingid":1224},{"bookingid":1533},{"bookingid":1336},{"bookingid":1071},{"bookingid":3244},{"bookingid":1316},{"bookingid":3464},{"bookingid":175},{"bookingid":384},{"bookingid":546},{"bookingid":2501},{"bookingid":2757},{"bookingid":158},{"bookingid":669},{"bookingid":2111},{"bookingid":32},{"bookingid":824},{"bookingid":3643},{"bookingid":2707},{"bookingid":1859},{"bookingid":2265},{"bookingid":3596},{"bookingid":1737},{"bookingid":3359},{"bookingid":2951},{"bookingid":2891},{"bookingid":2852},{"bookingid":1198},{"bookingid":3292},{"bookingid":1759},{"bookingid":3892},{"bookingid":474},{"

In [19]:
def get_guest_details(start_booking_id,end_booking_id):
  url = 'https://restful-booker.herokuapp.com/booking'
  booking_details = []
  for i in range(start_booking_id,end_booking_id+1):
    response = requests.get(f'{url}/{i}')
    booking_details.append(response.json())
  return booking_details

details = get_guest_details(1,9)
print(details)

[{'firstname': 'Eric', 'lastname': 'Wilson', 'totalprice': 648, 'depositpaid': True, 'bookingdates': {'checkin': '2021-05-04', 'checkout': '2021-09-03'}, 'additionalneeds': 'Breakfast'}, {'firstname': 'Eric', 'lastname': 'Brown', 'totalprice': 664, 'depositpaid': False, 'bookingdates': {'checkin': '2019-12-24', 'checkout': '2023-11-12'}, 'additionalneeds': 'Breakfast'}, {'firstname': 'Mark', 'lastname': 'Ericsson', 'totalprice': 105, 'depositpaid': True, 'bookingdates': {'checkin': '2016-05-23', 'checkout': '2018-08-19'}, 'additionalneeds': 'Breakfast'}, {'firstname': 'Eric', 'lastname': 'Jones', 'totalprice': 921, 'depositpaid': False, 'bookingdates': {'checkin': '2024-01-24', 'checkout': '2024-04-02'}}, {'firstname': 'Mary', 'lastname': 'Jackson', 'totalprice': 406, 'depositpaid': False, 'bookingdates': {'checkin': '2024-04-12', 'checkout': '2024-06-22'}}, {'firstname': 'Susan', 'lastname': 'Smith', 'totalprice': 930, 'depositpaid': True, 'bookingdates': {'checkin': '2018-11-25', 'ch

## ###Problem statement -->
You are tasked with developing a product management system that interacts with an API to add new products. Write a function that sends a POST request to the API endpoint for adding new products. The request payload should be in JSON format.

The API endpoint is available at: https://dummyjson.com/products/add
Input
The function will input product information and sends a POST request to the API endpoint.
### Output
    {'id': 195, 'title': 'BMW Pencil', 'price': 50, 'brand': 'BMW', 'category': 'Stationery'}

In [22]:
headers = { 'Content-Type': 'application/json' }
url = 'https://dummyjson.com/products/add'
body = {
    "title": "BMW Pencil",
    "price": 50,
    "brand": "BMW",
    "category": "Stationery"
}
string_data = json.dumps(body)
response = requests.post(url,headers = headers,data=string_data)
response.status_code

201

In [23]:
response.json()

{'id': 195,
 'title': 'BMW Pencil',
 'price': 50,
 'brand': 'BMW',
 'category': 'Stationery'}

PUT Method:
1. The PUT method is used to update existing data on the server.
2. It replaces the entire resource with the uddated data provided in the request
3.PUT requests are often used to update specific resources,such as modifying user information or updating a blog post

In [24]:
put_url = "https://dummyjson.com/products/1"
headers = { 'Content-Type': 'application/json' }
body = {'title': 'iPhone Galaxy +1'}
string_body = json.dumps(body)
put_response = requests.put(put_url,headers = headers , data = string_body)
put_response.status_code

200

In [25]:
put_response.json()

{'id': 1,
 'title': 'iPhone Galaxy +1',
 'price': 9.99,
 'discountPercentage': 7.17,
 'stock': 5,
 'rating': 4.94,
 'images': ['https://cdn.dummyjson.com/products/images/beauty/Essence%20Mascara%20Lash%20Princess/1.png'],
 'thumbnail': 'https://cdn.dummyjson.com/products/images/beauty/Essence%20Mascara%20Lash%20Princess/thumbnail.png',
 'description': 'The Essence Mascara Lash Princess is a popular mascara known for its volumizing and lengthening effects. Achieve dramatic lashes with this long-lasting and cruelty-free formula.',
 'brand': 'Essence',
 'category': 'beauty'}

DELETE Method:
1.The DELETE method is used to delete a specified resource on the server.
2. It removes the resource permanently from the server.
3. DELETE requests are used to remove data or resources from the server.


In [26]:
delete_response = requests.delete('https://dummyjson.com/products/2')
delete_response.status_code

200

In [27]:
delete_response.json()

{'id': 2,
 'title': 'Eyeshadow Palette with Mirror',
 'description': "The Eyeshadow Palette with Mirror offers a versatile range of eyeshadow shades for creating stunning eye looks. With a built-in mirror, it's convenient for on-the-go makeup application.",
 'category': 'beauty',
 'price': 19.99,
 'discountPercentage': 5.5,
 'rating': 3.28,
 'stock': 44,
 'tags': ['beauty', 'eyeshadow'],
 'brand': 'Glamour Beauty',
 'sku': 'MVCFH27F',
 'weight': 3,
 'dimensions': {'width': 12.42, 'height': 8.63, 'depth': 29.13},
 'warrantyInformation': '1 year warranty',
 'shippingInformation': 'Ships in 2 weeks',
 'availabilityStatus': 'In Stock',
 'reviews': [{'rating': 4,
   'comment': 'Very satisfied!',
   'date': '2024-05-23T08:56:21.618Z',
   'reviewerName': 'Liam Garcia',
   'reviewerEmail': 'liam.garcia@x.dummyjson.com'},
  {'rating': 1,
   'comment': 'Very disappointed!',
   'date': '2024-05-23T08:56:21.618Z',
   'reviewerName': 'Nora Russell',
   'reviewerEmail': 'nora.russell@x.dummyjson.com

In [30]:
url = 'https://dummyjson.com/recipes?limit=10&skip=0&select=name,image'
response = requests.get(url)
response.status_code

200

In [31]:
response.json()

{'recipes': [{'id': 1,
   'name': 'Classic Margherita Pizza',
   'image': 'https://cdn.dummyjson.com/recipe-images/1.webp'},
  {'id': 2,
   'name': 'Vegetarian Stir-Fry',
   'image': 'https://cdn.dummyjson.com/recipe-images/2.webp'},
  {'id': 3,
   'name': 'Chocolate Chip Cookies',
   'image': 'https://cdn.dummyjson.com/recipe-images/3.webp'},
  {'id': 4,
   'name': 'Chicken Alfredo Pasta',
   'image': 'https://cdn.dummyjson.com/recipe-images/4.webp'},
  {'id': 5,
   'name': 'Mango Salsa Chicken',
   'image': 'https://cdn.dummyjson.com/recipe-images/5.webp'},
  {'id': 6,
   'name': 'Quinoa Salad with Avocado',
   'image': 'https://cdn.dummyjson.com/recipe-images/6.webp'},
  {'id': 7,
   'name': 'Tomato Basil Bruschetta',
   'image': 'https://cdn.dummyjson.com/recipe-images/7.webp'},
  {'id': 8,
   'name': 'Beef and Broccoli Stir-Fry',
   'image': 'https://cdn.dummyjson.com/recipe-images/8.webp'},
  {'id': 9,
   'name': 'Caprese Salad',
   'image': 'https://cdn.dummyjson.com/recipe-imag

### API Status Codes
-1xx: indicate that request has been received and the server is countinuing to process it.

-2xx: indicate that the request was successful and the server has fulfilled it.

-3xx:indicate that further action needs to be taken to complete the requet(server is unable to process the request and diverting your request to different server).

-4xx:indicate that the client's request was invalid or cannot be processed by the server(e.g. endpoint is not available. basically due to client error).

-5xx:indicate that the server failed to fulfill a valid request(server does not have function to process your request).

# Inspect Response:
check the response status code to understand the result of the request
**error messanges:**extract and present detailed error messages to useres.
**conditional logic:** implement logic based on status codes for appropriate user feedback.
**retry mechanism:** create a mechanism to retry requests after a short delay fro transient errors.
**logging:**maintain logs to track error responses and improve your application.

1. 'https://dummyjson.com/products'
2. 'https://api.twitter.com/2/tweets/search/recent?query=from:elonmusk'
3. 'http://the-internet.herokuapp.com/status_codes/500'

In [32]:
response_200 = requests.get('https://dummyjson.com/products')
response_200.status_code

200

In [36]:
response_404 = requests.get('https://dummyjson.com/specific_item')
response_404.status_code

404

In [37]:
response_404.text

''

In [39]:
response_500 = requests.get('http://the-internet.herokuapp.com/status_codes/500')
response_500.status_code

500

In [42]:
print(response_500.text)

<!DOCTYPE html>
<!--[if IE 8]>         <html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->

<html>
  <head>
    <script src="/js/vendor/298279967.js"></script>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <title>The Internet</title>
    <link href="/css/app.css" rel="stylesheet">
    <link href="/css/font-awesome.css" rel="stylesheet">
    <script src="/js/vendor/jquery-1.11.3.min.js"></script>
    <script src="/js/vendor/jquery-ui-1.11.4/jquery-ui.js"></script>
    <script src="/js/foundation/foundation.js"></script>
    <script src="/js/foundation/foundation.alerts.js"></script>
    <script>
      $(document).foundation();
    </script>
  </head>
  <body>
    <div class="row">
      <div id="flash-messages" class="large-12 columns">
      
        
      
        
      
        
      
      </div>
    </div>
    <div class="row">
      <a href="https://github.com/to

In [43]:
response_auth = requests.get('https://api.twitter.com/2/tweets/search/recent?query=from:elonmusk')
response_auth.status_code

401

In [44]:
response_auth.json()

{'title': 'Unauthorized',
 'type': 'about:blank',
 'status': 401,
 'detail': 'Unauthorized'}

# REST API:
stands for Representational State Transer.
It is an architectural style that defines a set of constraints and principles for designing networked applications.
REST APIs are based on simplicity,scalability and statelessness.
they use standard HTTP methods like GET,POST,PUT,DELETE to perform operations on resources.
URLs(endpoints) are used to identify and access resources.
REST APIs primarily use JSON(JavaScript Object Notation) for data exchange.


# ##SOAP:
stands fro simple object access protocol
soap is a protocol for facilitating communication between applications over a network.
it uses XML as the data format for communication.